reference/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,315 @@
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+#include "core/reorder/rcm_kernels.hpp"
+
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The reordering namespace.
+ *
+ * @ingroup reorder
+ */
+namespace rcm {
+
+
+template <typename IndexType>
+void get_degree_of_nodes(std::shared_ptr<const ReferenceExecutor> exec,
+                         const size_type num_vertices,
+                         const IndexType *const row_ptrs,
+                         IndexType *const degrees)
+{
+    for (auto i = 0; i < num_vertices; ++i) {
+        degrees[i] = row_ptrs[i + 1] - row_ptrs[i];
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_RCM_GET_DEGREE_OF_NODES_KERNEL);
+
+
+/**
+ * Computes a level structure rooted at `root`, returning the height of the
+ * structure along with the index of the node of minimal degree in its last
+ * level.
+ */
+template <typename IndexType>
+std::pair<IndexType, size_type> rls_contender_and_height(
+    std::shared_ptr<const ReferenceExecutor> exec, const size_type num_vertices,
+    const IndexType root, const IndexType *const row_ptrs,
+    const IndexType *const col_idxs, const IndexType *const degrees)
+{
+    // This could actually be allocated in the calling scope, then reused.
+    Array<bool> visited_local(exec, num_vertices);
```

Comment:
Compiler doesn't like that _at all_. I believe this is because it apparently requires not an ordinary allocator, but a `std::_Bvector_base<_Alloc>::_Bit_alloc_type`, which there is no known conversion from `ExecutorAllocator<bool>` to.

Reply:
I'll take a look at it. Thanks!

Reply:
This seems to be an issue with the rebinding constructor (changing the value_type of an allocator). I thought I had taken care of that, but I'll investigate further

Reply:
Can you try removing the `explicit` from ExecutorAllocator's rebinding constructor
```
template <typename U>
explicit ExecutorAllocator(const ExecutorAllocator<U> &other)
```

Reply:
Compiles. Should I push that?

Reply:
Yes, please!